api/nakadi-event-bus-api.yaml

```diff
@@ -623,6 +539,349 @@ paths:
+              items:
+                description: list of subscriptions
+                type: array
+                items:
+                  $ref: '#/definitions/Subscription'
+            required:
+              - items
+              - _links
+
+  /subscriptions/{subscription_id}:
+    get:
+      tags:
+        - subscription-api
+      security:
+        - oauth2: ['nakadi.event_stream.read']
+      description: Returns a subscription identified by id.
+      parameters:
+        - $ref: '#/parameters/SubscriptionId'
+      responses:
+        '200':
+          description: OK
+          schema:
+            $ref: '#/definitions/Subscription'
+        '404':
+          description: Subscription not found
+          schema:
+            $ref: '#/definitions/Problem'
+
+  /subscriptions/{subscription_id}/cursors:
+    get:
+      tags:
+        - subscription-api
+      security:
+        - oauth2: ['nakadi.event_stream.read']
+      description: Exposes the currently committed offsets of a subscription.
+      parameters:
+        - $ref: '#/parameters/SubscriptionId'
+      responses:
+        '200':
+          description: Ok
+          schema:
+            type: object
+            properties:
+              items:
+                description: list of cursors for subscription
+                type: array
+                items:
+                  $ref: '#/definitions/Cursor'
+            required:
+              - items
+        '404':
+          description: Subscription not found
+          schema:
+            $ref: '#/definitions/Problem'
+    post:
+      tags:
+        - subscription-api
+      security:
+        - oauth2: ['nakadi.event_stream.read']
+      description: |
+        Endpoint for committing offsets of the subscription. The client must commit at least once
+        every 60 seconds, otherwise Nakadi will consider the client to be gone and will close the
+        connection.
+
+        When a batch is committed that also automatically commits all previous batches that were
+        sent in a stream for this partition.
+      parameters:
+        - name: subscription_id
+          in: path
+          type: string
+          description: Id of subscription
+          required: true
+        - name: X-Nakadi-StreamId
+          in: header
+          type: string
+          description: |
+            Id of stream which client uses to read events. It is not possible to make a commit for a terminated or
+            none-existing stream. Also the client can't commit something which was not sent to his stream.
+        - name: cursors
+          in: body
+          schema:
+            type: object
+            properties:
+              cursors:
+                description: |
+                  List of cursors that the consumer acknowledges to have successfully processed.
+                type: array
+                items:
+                  $ref: '#/definitions/Cursor'
+            required:
```

Comment:
> Since we can only add an optional field, due to backward compatibility, 

So sure, it can be _added_ in a compatible way. Let's suppose the topic indicator is `event_type_name` for a minute. Adding that field to the cursor isn't a compatibility break for any existing consumer clients today and the server can ignore if it sees it. 

But `event_type_name` can't be optional for multi-topic subscriptions on the write path. The server will have to return errors for cursors that don't have a topic indicator for a multi-type subscription. That's a breaking change in behaviour. In the future then, there'll have to be two modes on the subscription API, single request mode that doesn't require a topic indicator (to stop single mode clients crashing as soon as the feature is added) and multi-request mode which must have an `event_type_name`. The project code will end up with a logic switch that requires it check the subscription's shape and decide whether the cursor data is acceptable. There's also no way to articulate that variation in the API using the current Cursor definition even when it's extended; probably it's going to require a `SubscriptionUpdateCursor` to be captured clearly which can't be added later either because the definition is committed to re-using Cursor.

I think there's value in having one way to checkpoint instead of introducing complexity via modes later on that clients and the server have to upgrade to support.

One possible way around this, is to document clearly now, that the subscription consumer must submit the cursor it was given on the stream as is and with all the content supplied in that cursor; anything else is a 422. That allows you to extend the cursor later on. The problem with that is that it doesn't work well for swagger generated clients who are allowed to (or just will) drop data they don't understand and thus won't roundtrip the data back to the server. Consumers will have to have custom code for multi-subscription posting in that case. (this is similar to the definition for the business process category, where you can't use the generated client code.)

That's about as well as I can explain it :) I don't see a clean way to deal with this later on except to add complexity. On the other hand it can be designed for today.